Package.swift

```diff
@@ -4,23 +4,27 @@ import PackageDescription
 
 let package = Package(
     name: "llama",
+    platforms: [.macOS(.v11)],
```

Comment:
I haven't tried iOS but I do wonder if we are OK by limiting this package to macOS only. Can llamacpp run on iOS, tvOS or even watchOS @ggerganov?

Reply:
It can run even on a refrigerator 😄 

Jokes aside - I see no reason to limit this to just macOS

Reply:
i'm not sure why, but the compiler was not happy if platforms was not included, although that could be a metal thing?

Reply:
i don't have a way to test on watchOS or tvOS... i can set it to the minimum non-deprecated version and if someone tries to cross that bridge, we could update it?

Reply:
do you mind sharing the compiler error?

Reply:
![Uploading Screenshot 2023-09-08 at 11.16.04 AM.png…]()

Reply:
<img width="697" alt="Screenshot 2023-09-08 at 11 16 37 AM" src="https://github.com/ggerganov/llama.cpp/assets/62481661/8834ab90-ff81-494e-83ed-ae0ae282ba3e">

Reply:
too many errors, lol

Reply:
https://github.com/mybigday/llama.rn/blob/main/scripts/ggml-metal.m.patch

I've a patch for iOS/tvOS, for skip some unavailable APIs. I also set the minimum versions to 11: [podspec](https://github.com/mybigday/llama.rn/blob/main/llama-rn.podspec#L24).

The .metal file should be able to skip compile by add to resources, it should be loaded dynamically during runtime.

Reply:
Thats right. You shouldn’t be compiling the .metal file, it needs to be a copied resource for the spm target. That will cause your current code to access the bundle to have to change slightly since SPM generates a resources bundle with the metal file

Reply:
@jhen0409 , sorry do you mind giving me an ELI5? I'm a backend eng for the most part so not that familiar with Swift or React Native.

From what I can tell in the .m file, you're commenting out everything related to devices, but I'm not following why or if you're suggesting that I need to do something differently here.

RE: loading dynamically during runtime, when I imported llama through the Swift Package Manager, the .metallib was in the llama bundle. I think that the other part of the code (the else-clause of the GGML_SWIFT) is trying to load the file from source, but I'm not sure if that's necessary for Swift at least. Is there a reason to do that?

Reply:
The API like MTLCopyAllDevices and some property in MTLDevice is not supported or requires newer version on iOS. I can open a PR and put more info for that.

> RE: loading dynamically during runtime, when I imported llama through the Swift Package Manager, the .metallib was in the llama bundle. I think that the other part of the code (the else-clause of the GGML_SWIFT) is trying to load the file from source, but I'm not sure if that's necessary for Swift at least. Is there a reason to do that?

I'm not sure what default.metallib used for. You can try use `resources: [ .copy(".metal") ]` in target & use `SWIFTPM_MODULE_BUNDLE` instead.

Reply:
Hey folks, I would appreciate help on this. I'm seeing in my build logs that it's compiling the .metal file even if I put it in `resources`.

For example, if I do:
```
#if arch(arm) || arch(arm64)
let platforms: [SupportedPlatform]? = [
    .macOS(.v11),
    .iOS(.v11),
    .watchOS(.v4),
    .tvOS(.v11)
]
let exclude: [String] = []
let resources: [Resource]? = [
    .copy("ggml-metal.metal"),
    .copy("README.md")  // just to validate that files get copied
]
let additionalSources: [String] = ["ggml-metal.m"]
let additionalSettings: [CSetting] = [
    .unsafeFlags(["-fno-objc-arc"]),
    .define("GGML_SWIFT"),
    .define("GGML_USE_METAL")
]
#else
```

I still see the default.metallib in my resources:
```
ls /Users/.../Library/Developer/Xcode/DerivedData/.../Build/Products/Debug/....app/Contents/Resources/llama_llama.bundle/Contents/Resources/
README.md		default.metallib
```


<img width="519" alt="Screenshot 2023-09-08 at 6 37 22 PM" src="https://github.com/ggerganov/llama.cpp/assets/62481661/e840a545-19e1-41a3-a75e-c8916069a987">


Could it be because the file is in the project root and the target path is "."?

Reply:
what if you exclude ggml-metal.metal like it is done in master right now, but then add it in the resources section. That should do it I think

Reply:
https://developer.apple.com/documentation/packagedescription/target/exclude#discussion

I think it doesn't work because exclude takes precedence over resources. For example, I pushed a branch https://github.com/ggerganov/llama.cpp/pull/3091/files, and my build logs don't show the .metal file:

<img width="493" alt="Screenshot 2023-09-08 at 7 57 54 PM" src="https://github.com/ggerganov/llama.cpp/assets/62481661/0eac3a08-0b7a-4c30-82ed-9e49cb03cd3a">

Reply:
https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/MTLBestPracticesGuide/FunctionsandLibraries.html

Just a sanity check, but this documentation is saying that .metal files get automatically compiled, and that seems to be the case from what I've tried. Is there a specific reason why we need to compile from source?

Reply:
> https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/MTLBestPracticesGuide/FunctionsandLibraries.html
> 
> Just a sanity check, but this documentation is saying that .metal files get automatically compiled, and that seems to be the case from what I've tried.

Got it, it looks like a better way if we use it on Xcode.